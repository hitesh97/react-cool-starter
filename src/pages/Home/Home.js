/* @flow */

import React, { PureComponent } from 'react';
import { compose } from 'redux';
import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';
import Helmet from 'react-helmet';

import { featuredProductsAction } from '../../actions';
import type { Home as HomeType, Dispatch, ReduxState } from '../../types';
import { FeaturedProductsList } from '../../components';
import styles from './styles.scss';

type Props = {
  home: HomeType,
  fetchFeaturedProductsIfNeeded: () => void
};

// Export this for unit testing more easily
export class Home extends PureComponent<Props> {
  componentDidMount() {
    const { fetchFeaturedProductsIfNeeded } = this.props;

    fetchFeaturedProductsIfNeeded();
  }

  renderProductsList = () => {
    const { home } = this.props;

    if (
      !home.readyStatus ||
      home.readyStatus === 'FEATURED_PRODUCTS_REQUESTING' ||
      home.readyStatus === 'FEATURED_PRODUCTS_INVALID'
    ) {
      return <p>Loading products...</p>;
    }

    if (home.readyStatus === 'FEATURED_PRODUCTS_FAILURE') {
      return (
        <p>
          Oops, Failed to load Featured Products!
          <br />
          {home.err}
        </p>
      );
    }
    return <FeaturedProductsList list={home.featuredProducts} />;
  };

  render() {
    return (
      <div className={styles.Home}>
        <Helmet title="Home" />
        {this.renderProductsList()}
      </div>
    );
  }
}

const connector = connect(
  ({ home }: ReduxState) => ({ home }),
  (dispatch: Dispatch) => ({
    fetchFeaturedProductsIfNeeded: () =>
      dispatch(featuredProductsAction.fetchFeaturedProductsIfNeeded())
  })
);

export default compose(
  withRouter,
  connector
)(Home);
