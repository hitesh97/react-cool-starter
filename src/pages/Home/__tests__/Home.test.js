import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import { MemoryRouter } from 'react-router-dom';

import { Home } from '../Home';

describe('<Home />', () => {
  const tree = (props, actions) =>
    renderer
      .create(
        <MemoryRouter>
          <Home {...props} {...actions} />
        </MemoryRouter>
      )
      .toJSON();

  it('should call fetchFeaturedProductsIfNeeded when componentDidMount', () => {
    const mockAction = jest.fn();
    const props = {
      home: {}
    };
    const actions = {
      fetchFeaturedProductsIfNeeded: mockAction
    };

    mount(
      <MemoryRouter>
        <Home {...props} {...actions} />
      </MemoryRouter>
    );

    expect(mockAction).toHaveBeenCalled();
  });
  /* 
  it('renders the loading status if data invalid', () => {
    const props = {
      home: { readyStatus: 'USERS_INVALID' }
    };
    const actions = { fetchUsersIfNeeded: () => {} };

    expect(tree(props, actions)).toMatchSnapshot();
  });

  it('renders the loading status if requesting data', () => {
    const props = {
      home: { readyStatus: 'USERS_REQUESTING' }
    };
    const actions = { fetchUsersIfNeeded: () => {} };

    expect(tree(props, actions)).toMatchSnapshot();
  });

  it('renders an error if loading failed', () => {
    const props = {
      home: { readyStatus: 'USERS_FAILURE' }
    };
    const actions = { fetchUsersIfNeeded: () => {} };

    expect(tree(props, actions)).toMatchSnapshot();
  });

   it('renders the <UserList /> if loading was successful', () => {
    const props = {
      home: {
        readyStatus: 'USERS_SUCCESS',
        list: [{ id: '1', name: 'Welly' }]
      }
    };
    const actions = { fetchUsersIfNeeded: () => {} };

    expect(tree(props, actions)).toMatchSnapshot();
  }); */

  it('renders the <FeaturedProductsList /> if loading was successful', () => {
    const props = {
      home: {
        readyStatus: 'FEATURED_PRODUCTS_SUCCESS',
        featuredProducts: [
          [
            {
              imageName: 'FTB-1-2HR.jpg',
              imgUrl: '/Images/HomeFeaturedProducts/FTB-1-2HR.jpg',
              title: 'Exfo FTB-1 Familie',
              subtitle:
                'Het FTB-1 Platform is een open testoplossing voor het.',
              links: [
                {
                  linkText: 'Meer informatie',
                  linkUrl:
                    '/products/Exfo/SDH-SONET-Ethernet/FTB-1?BaseModelId=89542',
                  linkType: 'button'
                },
                {
                  linkText: 'Alles weergeven SDH / SONET / Ethernet',
                  linkUrl: '/product-group/5/sdh-sonet-ethernet',
                  linkType: 'href'
                }
              ]
            }
          ]
        ]
      }
    };
    const actions = { fetchFeaturedProductsIfNeeded: () => {} };

    expect(tree(props, actions)).toMatchSnapshot();
  });
});
