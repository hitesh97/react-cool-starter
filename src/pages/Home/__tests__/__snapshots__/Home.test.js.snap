// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Home /> renders the <FeaturedProductsList /> if loading was successful 1`] = `
<div
  className="Home"
>
  <div
    className="FeaturedProductsList"
  >
    <h4>
      Featured PRoducts
    </h4>
    <ul>
      <li>
        <img
          src="https://d3fdwrtpsinh7j.cloudfront.netundefined"
        />
        <h1 />
        <h4 />
      </li>
    </ul>
  </div>
</div>
`;
